Simulating multiple sequences without gBGC: [R/Simulation.R](https://github.com/xyz111131/PhyloAcc/blob/master/R/Simulation.R) generated the simulated sequences for ratite phylogeny for different accelerated scenarios. The resulting files are *Simulation_ratite/simu_500_200_diffr_** *. Fasta is concatenated alignment file for all simulated elements under each scenario; Bed is the coordinate file.
1-1 is for "all branches conserved"; 2-* is for 1) only kiwi clade accelerated; 2) only emu/cassowary branches accelerated; 3) only rheas clade accelerated; 4) only ostrich accelerated; 5) all ratites accelerated except ostrich and moa; 6) all ratites accelerated; 7) both ratites and volant tinamous accelerated; 8) 5 random species across all birds accelerated. These files are used to generate results in the PhyloAcc manuscript.

Simulating DNA sequences with gBGC: [R/Simulation_gBGC.R](https://github.com/xyz111131/PhyloAcc/blob/master/R/Simulation_gBGC.R). In this R script, we fixed gBGC only affecting emu and cassowary branches (and their common ancestor) within ratites as well as finch, flycatcher and swift within neognath birds, and altered selection patterns: 0) all branches are conserved; 1) only kiwi clade accelerated; 2) only emu/cassowary branches accelerated; 3) only rheas clade accelerated; 4) only ostrich accelerated; 5) all ratites accelerated except ostrich and moa; 6) all ratites accelerated. The resulting files are in *V2_GBGC/Simulation/*. Bed file is shared in all scenarios.
